Replace IOContexts with execution env and contexts

hubertp opened the pull request:
### Pull Request Description

As per design, IOContexts controlled via type signatures are going away. They are replaced by explicit `Context.if_enabled` runtime checks that will be added to particular method implementations.

`production`/`development` `IOPermissions` are replaced with `live` and `design` execution enviornment. Currently, the `live` env has a hardcoded list of allowed contexts i.e. `Input` and `Output`.

### Important Notes

As per design PR-55. Closes #6129. Closes #6131.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ x All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] ~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~

hubertp committed:
Replace IOContexts with execution env and contexts

As per design, IOContexts controlled via type signatures are going away.
They are replaced by explicit `Context.if_enabled` runtime checks that
will be added to particular method implementations.

`production`/`development` `IOPermissions` are replaced with `live` and
`design` execution enviornment. Currently, the `live` env has a
hardcoded list of allowed contexts i.e. `Input` and `Output`.

hubertp committed:
changelog

hubertp committed:
make native image build happy

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Errors/Common.enso:
```diff
@@ -231,3 +231,21 @@ type No_Such_Conversion
+@Builtin_Type
+type Forbidden_Operation
+    ## PRIVATE
+       An error that occurs when the action is not allowed to perform the operation in the given context
```
```suggestion
       An error that occurs when the action is not allowed to perform the operation in the given context.
```

Nitpick, but let's finish sentences in docs with dots.

hubertp replied:
We sometimes do, sometimes don't. I really believe we should have a linter for that.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/state/ExecutionEnvironment.java:
```diff
@@ -0,0 +1,77 @@
+      case "design":
+        return DESIGN;
+      default:
+        return new ExecutionEnvironment(name);
```
Does it make sense to return a newly-constructed environment for a name that is not `live` nor `design`? Such environment will always be equivalent to `design` anyway, since there is no way to customize its permissions.

radeusgd replied:
Intuitively I'd guess that accessing an environment other than these two, at least in the current setup would indicate a bug.

hubertp replied:
Throwing an exception now

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/state/ExecutionEnvironment.java:
```diff
@@ -0,0 +1,77 @@
+
+  private static final ExecutionEnvironment initLive() {
+    Map<String, Boolean> contexts = new HashMap<>();
+    contexts.put("Input", true);
```
Shouldn't these names be constants?

hubertp replied:
Ideally we would get those from the builtin's constructor list except... that won't be initialized yet.
I can make it slightly less hacky, though.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/runtime/ContextName.java:
```diff
@@ -0,0 +1,18 @@
+
+@BuiltinMethod(type = "Context", name = "name", description = "Name of the context.")
+public class ContextName extends Node {
+  Object execute(Atom self) {
```
Does it make sense to have this as a builtin? Simple:
```haskell
case self of
    Input -> "Input"
    Output -> "Output"
```
might be simpler.

hubertp replied:
I don't like hardcoding stuff, but sure.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java:
```diff
@@ -226,6 +229,10 @@ public ForbiddenOperation getForbiddenOperation() {
     return forbiddenOperation;
   }
 
+  public Atom makeNotImplemented(String operation) {
```
Why are we introducing `NotImplemented` type?

hubertp replied:
This was requested by @jdunkerley to deal with a situation when someone does a mismatch in environments. That behaviour is currently undefined really.

hubertp replied:
Realized there is `Unimplemted` so I changed it to that.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/state/ExecutionEnvironment.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  @CompilerDirectives.TruffleBoundary
+  public ExecutionEnvironment withContextEnabled(Atom context) {
```
Shouldn't we strive for making this function working on fast path?

hubertp replied:
I would have to refactor pretty much any call to map i.e. almost every line in that method, to make native image build happy. I would say that kind of defeats the purpose?

JaroslavTulach replied:
What's your purpose? Is the purpose to run Enso fast? Well, in such case that's not happening as soon as somebody uses `withContextEnabled` - that's always going to be slow.

> refactor pretty much any call to map

Possibly using `HashMap` for storage of two values (`Input`, `Output`) wasn't good idea to begin with?

hubertp replied:
Fine, replaced by two arrays that emulate the map. The reason why I don't hardcode those in fields is that I'm trying to be generic enough in preparation for potential future `Context` values.

hubertp replied:
It is also a big overkill given that **currently** we only have `Input` and `Output`. So if you feel strongly about it, I could just go back to storing those in fields.

JaroslavTulach replied:
I am assuming that `withContextEnabled`/`withContextDisabled` operations are going to be spread all around the codebase possibly included inside of `.map` lambdas processing large data structures and we *want to make them fast* in order to avoid slowing down the rest of the interpreter. Is such assumption a false expectation? As I am seeing a disagreement and I don't understand what are its roots...

> replaced by two arrays that emulate the map

The code is now less readable, but at least it doesn't need to be behind `@TruffleBoundary`. However I cannot judge how well it partially evaluates without looking at IGV graphs. Btw. what do we want to see in IGV when we look at `withContextEnabled`?
- when already enabled - a single field/bit check to verify the context is already enabled
- when disabled - more work - field write and switching the `State` to a new configuration (possibly also a costly allocation of the new configuration) 

> trying to be generic enough in preparation for potential future Context values.

Generic code can't be as fast as code that handles/speculates on fixed set of values. 

> ... if **you feel** strongly ...

Maybe we could change one of our benchmarks to use `withContextEnabled` few million of times and compare it to a benchmark without context manipulation? Then **we all** would have something measurable to know how to **feel** about this matter.

hubertp replied:
I will file a separate ticket to investigate if this becomes a perf issue.

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/ValuesGenerator.java:
```diff
@@ -721,8 +721,10 @@ public List<Value> allValues() throws Exception {
+
       if (m.getName().startsWith("type")) {
         if (m.getReturnType() == Value.class) {
+          System.out.println("All methhods " + getClass() + " for " + m);
```
Accidental commit, I guess.

JaroslavTulach commented on engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/ContextTest.scala:
```diff
@@ -0,0 +1,217 @@
+  }
+}
+
+class LiveExecutionEnvironmentTest extends InterpreterTest {
```
Two classes in a single file. That's unusual, I'd say.

hubertp replied:
Yeah, it has been the same way with `IOContext` so I just stick to it. I can split it if you want.

hubertp committed:
Add cli option to control environment

Added a fix for #6131 to allow for some testing.

hubertp committed:
PR review

hubertp commented:
@jdunkerley added a CLI option `--execution-environment`, as requested. Part of #6131.

hubertp committed:
PR review

hubertp committed:
revert to original truffleboundary

JaroslavTulach commented on engine/runner/src/main/scala/org/enso/runner/Main.scala:
```diff
@@ -355,6 +356,16 @@ object Main {
+      .numberOfArgs(1)
+      .argName("name")
+      .desc(
+        "Execution environment to use during execution (`live` or `design`)"
```
What is the default? `design`, I think, but it shall be part of the message, I believe.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/runtime/Context.java:
```diff
@@ -0,0 +1,27 @@
+    return List.of(new Cons(INPUT_CONTEXT), new Cons(OUTPUT_CONTEXT));
+  }
+
+  public static String INPUT_CONTEXT = "Input";
```
Missing `final`. Name of the field reads as `Context.INPUT_CONTEXT` from an external class usage - isn't the "context" duplicated?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/runtime/ContextIsEnabledNode.java:
```diff
@@ -0,0 +1,32 @@
+              .getBuiltins()
+              .error()
+              .makeUnimplemnted("execution environment mismatch");
+      throw new PanicException(error, this);
```
I really doubt a `Panic` is what we want! I don't even understand what's the purpose of specifying environment name in the `is_enabled` check. Environments are supposed to be a set of contexts. All the user wants is to take a context and check if it is currently enabled. E.g. `Context.Input.is_enabled`.

hubertp replied:
This isn't really specified well in the design doc. Nor is the usage of the environment name so if you feel strongly about it, probably worth bringing up with that doc's authors.

jdunkerley replied:
At least for the moment - there is no intention to specify anything other than the current context when calling this function. 

If we find a need to use this we can revisit.

JaroslavTulach replied:
> At least for the moment - there is no intention to specify anything other than the current context when calling this function.

Obviously! Specifying anything else makes no sense. At least I cannot envision the use-case.

I'd say: One doesn't care about the environment. One only wants to find out if a `Context` `is_enabled` "right now". Nothing else makes sense.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/runtime/ContextIsEnabledNode.java:
```diff
@@ -0,0 +1,32 @@
+          EnsoContext.get(this)
+              .getBuiltins()
+              .error()
+              .makeUnimplemnted("execution environment mismatch");
```
Typo in `makeUnimplemnted`

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/runtime/RuntimeCurrentExecutionEnvironmentNode.java:
```diff
@@ -0,0 +1,17 @@
+
+@BuiltinMethod(
+    type = "Runtime",
+    name = "current_execution_environment",
```
This method shall not exists from my API designer perspective as it bypasses the `Context`. I assume @jdunkerley wants the method to do production/non-production property configuration, but we should offer other means for that - not this method.

hubertp replied:
We agreed that `Execution_Environment` is really an alias for `Text` and I believe I'm following the design doc so I don't really understand the comment.

jdunkerley replied:
I won't be using this for configuration - from the libs perspective I don't expect this to be used.

JaroslavTulach replied:
There is a story about [magical strings](http://wiki.apidesign.org/wiki/MagicalStrings). I have a feeling the returned textual value has the potential to provide "magical" insights into something that should rather be hidden. 

> ...I don't expect this to be used.

Personally I'd rather not expose such API, especially if there is no use-case for it.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java:
```diff
@@ -226,6 +229,10 @@ public ForbiddenOperation getForbiddenOperation() {
     return forbiddenOperation;
   }
 
+  public Atom makeUnimplemnted(String operation) {
```
Typo in method name.

hubertp committed:
Address more PR comments

Changed the default execution env to `live` since that is probably what
we want for CLI anyway.
Got rid of `TruffleBoundary` annotations by using straight arrays that
emulate map.

jdunkerley commented on engine/runner/src/main/scala/org/enso/runner/Main.scala:
```diff
@@ -355,6 +356,16 @@ object Main {
+      .numberOfArgs(1)
+      .argName("name")
+      .desc(
+        "Execution environment to use during execution (`live`/`design`). Defaults to: `live`."
```
should default to "design" I believe...

hubertp replied:
I had it like that but figured we want to have `live` for CLI. Will revert.

jdunkerley commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/state/State.java:
```diff
@@ -2,35 +2,46 @@
-  public State withInputAllowedIn(String name) {
-    return new State(container, ioPermissions.allowInputIn(name));
+  public State withContextEnabledIn(Atom context, String environemtnName) {
+    if (executionEnvironment.getName().equals(environemtnName)) {
```
```suggestion
  public State withContextEnabledIn(Atom context, String environmentName) {
    if (executionEnvironment.getName().equals(environmentName)) {
```

jdunkerley commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/state/State.java:
```diff
@@ -2,35 +2,46 @@
-  public State withOutputAllowedIn(String name) {
-    return new State(container, ioPermissions.allowOutputIn(name));
+  public State withContextDisabledIn(Atom context, String environemtnName) {
+    if (executionEnvironment.getName().equals(environemtnName)) {
```
```suggestion
  public State withContextDisabledIn(Atom context, String environmentName) {
    if (executionEnvironment.getName().equals(environmentName)) {
```

jdunkerley reviewed (approved):
Couple of nit pics

4e6 commented on engine/runner/src/main/scala/org/enso/runner/ContextFactory.scala:
```diff
@@ -40,8 +40,12 @@ class ContextFactory {
     strictErrors: Boolean                  = false,
     useGlobalIrCacheLocation: Boolean      = true,
     enableAutoParallelism: Boolean         = false,
+    executionEnvironment: Option[String]   = None,
```
Please add `@param` to the method doc

4e6 commented on engine/runner/src/main/scala/org/enso/runner/Main.scala:
```diff
@@ -540,7 +552,8 @@ object Main {
     inspect: Boolean,
-    dump: Boolean
+    dump: Boolean,
+    executionEnvironment: Option[String]
```
Please add `@param` to the method doc

hubertp committed:
Address PR comments

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/6129-runtime-context

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/6129-runtime-context
